commit 9b4fb5cec031f81ef436bf2cfd9fc265e25f6e45
Author: Will Deacon <will@kernel.org>
Date:   Thu Dec 19 17:40:21 2019 +0000

    fault_inject: Don't rely on "return value" from WRITE_ONCE()
    
    It's a bit weird that WRITE_ONCE() evaluates to the value it stores and
    it's different to smp_store_release(), which can't be used this way.
    
    In preparation for preventing this in WRITE_ONCE(), change the fault
    injection code to use a local variable instead.
    
    Cc: Akinobu Mita <akinobu.mita@gmail.com>
    Signed-off-by: Will Deacon <will@kernel.org>

diff --git a/lib/fault-inject.c b/lib/fault-inject.c
index 8186ca84910b..ce12621b4275 100644
--- a/lib/fault-inject.c
+++ b/lib/fault-inject.c
@@ -106,7 +106,9 @@ bool should_fail(struct fault_attr *attr, ssize_t size)
 		unsigned int fail_nth = READ_ONCE(current->fail_nth);
 
 		if (fail_nth) {
-			if (!WRITE_ONCE(current->fail_nth, fail_nth - 1))
+			fail_nth--;
+			WRITE_ONCE(current->fail_nth, fail_nth);
+			if (!fail_nth)
 				goto fail;
 
 			return false;